File: native/python/device_flow.py
import base64
import json
import os
import sys
import time
import requests

VOUCH_ISSUER = os.environ.get('VOUCH_ISSUER', 'https://us.vouch.sh')
CLIENT_ID = os.environ.get('VOUCH_CLIENT_ID')

if not CLIENT_ID:
    print('Error: VOUCH_CLIENT_ID environment variable is required')
    sys.exit(1)


def decode_access_token(token):
    """Hardware claims are in the access token JWT (RFC 9068), not the id_token."""
    payload = token.split('.')[1]
    payload += '=' * (4 - len(payload) % 4)
    return json.loads(base64.urlsafe_b64decode(payload))


def fetch_userinfo(access_token):
    resp = requests.get(
        f'{VOUCH_ISSUER}/oauth/userinfo',
        headers={'Authorization': f'Bearer {access_token}'},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()


# Step 1: Request device code
response = requests.post(
    f'{VOUCH_ISSUER}/oauth/device',
    data={
        'client_id': CLIENT_ID,
        'scope': 'openid email',
    },
)
response.raise_for_status()
device_data = response.json()

# Step 2: Display instructions to user
print(f"\nTo sign in, visit: {device_data['verification_uri']}")
print(f"Enter code: {device_data['user_code']}\n")

# Step 3: Poll for token
interval = device_data.get('interval', 5)
while True:
    time.sleep(interval)

    token_response = requests.post(
        f'{VOUCH_ISSUER}/oauth/token',
        data={
            'grant_type': 'urn:ietf:params:oauth:grant-type:device_code',
            'device_code': device_data['device_code'],
            'client_id': CLIENT_ID,
        },
    )

    if token_response.status_code == 200:
        tokens = token_response.json()
        print("Authenticated!")
        print(f"Access token: {tokens['access_token'][:20]}...")

        # Step 4: Fetch user info and decode hardware claims from access token
        userinfo = fetch_userinfo(tokens['access_token'])
        at_claims = decode_access_token(tokens['access_token'])
        print(f"Email: {userinfo.get('email', 'N/A')}")
        print(f"Hardware verified: {at_claims.get('hardware_verified', False)}")
        if at_claims.get('hardware_aaguid'):
            print(f"Hardware AAGUID: {at_claims['hardware_aaguid']}")

        # Step 5: Demonstrate post-auth API call with the access token
        print("\n--- Post-auth API call ---")
        userinfo2 = fetch_userinfo(tokens['access_token'])
        print(f"Second userinfo call succeeded: {userinfo2.get('email')}")

        break

    error = token_response.json().get('error')
    if error == 'authorization_pending':
        continue
    elif error == 'slow_down':
        interval += 5
    elif error == 'expired_token':
        print('Device code expired. Please try again.')
        sys.exit(1)
    elif error == 'access_denied':
        print('Access denied by user.')
        sys.exit(1)
    else:
        print(f'Error: {token_response.json()}')
        sys.exit(1)

File: native/node/device_flow.js
import { createPublicKey, verify as verifySignature } from 'node:crypto';

const VOUCH_ISSUER = process.env.VOUCH_ISSUER || 'https://us.vouch.sh';
const CLIENT_ID = process.env.VOUCH_CLIENT_ID;

/**
 * Verify a Vouch access token against the issuer's published JWKS.
 *
 * hardware_verified is only in the access token, not the id_token. The access token is
 * an ES256-signed RFC 9068 JWT, so verify it rather than decoding the payload -- an
 * agent that acts on an unverified claim is acting on whatever it was handed.
 *
 * Uses only node:crypto so this example keeps its zero-dependency footprint; a real
 * agent would reach for a JOSE library and cache the JWKS.
 */
async function verifyAccessToken(token) {
  const [rawHeader, rawPayload, rawSignature] = token.split('.');
  if (!rawSignature) throw new Error('access token is not a JWS');

  const header = JSON.parse(Buffer.from(rawHeader, 'base64url').toString());
  // RFC 9068 access tokens carry typ: at+jwt. Requiring it rejects id_tokens, which
  // are not bearer credentials.
  if (header.typ?.toLowerCase() !== 'at+jwt') {
    throw new Error(`unexpected token typ: ${header.typ}`);
  }

  const jwks = await (await fetch(`${VOUCH_ISSUER}/oauth/jwks`)).json();
  const jwk = jwks.keys.find((k) => k.kid === header.kid);
  if (!jwk) throw new Error(`kid ${header.kid} not published in JWKS`);

  const key = createPublicKey({ key: jwk, format: 'jwk' });
  const data = Buffer.from(`${rawHeader}.${rawPayload}`);
  const signature = Buffer.from(rawSignature, 'base64url');
  // JOSE encodes ECDSA signatures as raw r||s rather than DER.
  const ok = jwk.kty === 'EC'
    ? verifySignature('sha256', data, { key, dsaEncoding: 'ieee-p1363' }, signature)
    : verifySignature('sha256', data, key, signature);
  if (!ok) throw new Error('access token signature did not verify');

  const claims = JSON.parse(Buffer.from(rawPayload, 'base64url').toString());
  if (claims.iss !== VOUCH_ISSUER) throw new Error('issuer mismatch');
  if (![claims.aud].flat().includes(CLIENT_ID)) throw new Error('audience mismatch');
  if (claims.exp * 1000 < Date.now()) throw new Error('access token expired');
  return claims;
}

if (!CLIENT_ID) {
  console.error('Error: VOUCH_CLIENT_ID environment variable is required');
  process.exit(1);
}

async function fetchUserInfo(accessToken) {
  const response = await fetch(`${VOUCH_ISSUER}/oauth/userinfo`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  if (!response.ok) {
    throw new Error(`UserInfo request failed: ${response.status}`);
  }
  return response.json();
}

async function deviceFlow() {
  // Step 1: Request device code
  const deviceResponse = await fetch(`${VOUCH_ISSUER}/oauth/device`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      client_id: CLIENT_ID,
      scope: 'openid email',
    }),
  });

  if (!deviceResponse.ok) {
    throw new Error(`Device request failed: ${deviceResponse.status}`);
  }

  const deviceData = await deviceResponse.json();

  // Step 2: Display instructions to user
  console.log(`\nTo sign in, visit: ${deviceData.verification_uri}`);
  console.log(`Enter code: ${deviceData.user_code}\n`);

  // Step 3: Poll for token
  let interval = (deviceData.interval || 5) * 1000;

  while (true) {
    await new Promise((resolve) => setTimeout(resolve, interval));

    const tokenResponse = await fetch(`${VOUCH_ISSUER}/oauth/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: deviceData.device_code,
        client_id: CLIENT_ID,
      }),
    });

    if (tokenResponse.ok) {
      const tokens = await tokenResponse.json();
      console.log('Authenticated!');
      console.log(`Access token: ${tokens.access_token.slice(0, 20)}...`);

      // Step 4: Fetch user info and verify the access token's hardware claims
      const userInfo = await fetchUserInfo(tokens.access_token);
      const atClaims = await verifyAccessToken(tokens.access_token);
      console.log(`Email: ${userInfo.email || 'N/A'}`);
      console.log(`Hardware verified: ${atClaims.hardware_verified || false}`);
      console.log(`acr: ${atClaims.acr || 'N/A'}`);
      console.log(`amr: ${(atClaims.amr || []).join(', ') || 'N/A'}`);

      // Step 5: Demonstrate post-auth API call with the access token
      console.log('\n--- Post-auth API call ---');
      const userInfo2 = await fetchUserInfo(tokens.access_token);
      console.log(`Second userinfo call succeeded: ${userInfo2.email}`);

      return;
    }

    const { error } = await tokenResponse.json();
    switch (error) {
      case 'authorization_pending':
        continue;
      case 'slow_down':
        interval += 5000;
        continue;
      case 'expired_token':
        throw new Error('Device code expired. Please try again.');
      case 'access_denied':
        throw new Error('Access denied by user.');
      default:
        throw new Error(`Unexpected error: ${error}`);
    }
  }
}

deviceFlow().catch((err) => {
  console.error(`Error: ${err.message}`);
  process.exit(1);
});
